src/Sylius/Bundle/CartBundle/spec/Provider/CartProviderSpec.php

```diff
@@ -30,79 +33,100 @@ function let(
     ) {
-        $this->beConstructedWith($context, $manager, $cartFactory, $repository, $eventDispatcher);
+        $this->beConstructedWith($context, $manager, $cartFactory, $cartRepository, $eventDispatcher);
     }
 
     function it_is_initializable()
     {
-        $this->shouldHaveType('Sylius\Bundle\CartBundle\Provider\CartProvider');
+        $this->shouldHaveType(CartProvider::class);
     }
 
     function it_implements_Sylius_cart_provider_interface()
     {
-        $this->shouldImplement('Sylius\Component\Cart\Provider\CartProviderInterface');
+        $this->shouldImplement(CartProviderInterface::class);
     }
 
     function it_looks_for_cart_by_identifier_if_any_in_storage(
-        $context,
-        $repository,
-        $eventDispatcher,
+        CartContextInterface $context,
+        RepositoryInterface $cartRepository,
+        EventDispatcherInterface $eventDispatcher,
         CartInterface $cart
     ) {
         $context->getCurrentCartIdentifier()->willReturn(3);
-        $repository->find(3)->willReturn($cart);
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_INITIALIZE, Argument::any())->shouldNotBeCalled();
+        $cartRepository->find(3)->willReturn($cart);
+        $eventDispatcher->dispatch(
+            SyliusCartEvents::CART_INITIALIZE,
+            Argument::type(GenericEvent::class)
+        )->shouldBeCalled();
+
+        $context->setCurrentCartIdentifier($cart)->shouldNotBeCalled();
 
         $this->getCart()->shouldReturn($cart);
     }
 
     function it_creates_new_cart_if_there_is_no_identifier_in_storage(
-        $context,
-        $repository,
+        CartContextInterface $context,
+        RepositoryInterface $cartRepository,
+        EventDispatcherInterface $eventDispatcher,
+        CartInterface $cart,
         FactoryInterface $cartFactory,
-        $eventDispatcher,
         CartInterface $cart
     ) {
         $context->getCurrentCartIdentifier()->willReturn(null);
         $cartFactory->createNew()->willReturn($cart);
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_INITIALIZE, Argument::any())->shouldBeCalled();
+        $cartRepository->find()->shouldNotBeCalled();
+
+        $this->initialize_cart_should_be_called_on($eventDispatcher);
+
+        $context->setCurrentCartIdentifier($cart)->shouldBeCalled();
 
         $this->getCart()->shouldReturn($cart);
     }
 
     function it_creates_new_cart_if_identifier_is_wrong(
-        $context,
+        CartContextInterface $context,
+        RepositoryInterface $cartRepository,
+        EventDispatcherInterface $eventDispatcher,
+        CartInterface $cart,
         FactoryInterface $cartFactory,
-        $repository,
-        $eventDispatcher,
         CartInterface $cart
     ) {
         $context->getCurrentCartIdentifier()->willReturn(7);
-        $repository->find(7)->shouldBeCalled()->willReturn(null);
+        $cartRepository->find(7)->shouldBeCalled()->willReturn(null);
```

Comment:
the createNew() on FactoryInterface should be called only if there is no cart in RepositoryInterface. It's good to know that in this case we are actually calling find() on RepositoryInterface. Behaviour of this method is to actually check in the RepositoryInterface if the cart is there